### DISCUSSION TITLE
masks

### DISCUSSION THREAD

**Discussion by amylazicki** [HEXRD/hexrdgui#1397] [Q&A]
I am running hexrd on a PC and installed the most recent stable version.  I played around with masking out features and found it very nonintuitive.  In hand-drawn masks:  what does right-clicking do?... what do you mean by a new line?  Nothing seems to happen for me with a right click.  I can outline a region with left clicks and it's successfully masked (most of the time... sometimes for some reason I can't find, both plates just disappear and then I just have to open the mask management menu and delete that mask).  Sometimes the left clicks appear as blue dots, sometimes as orange or green or red... what does this mean?  I would sort of like to be able to see the outline of the masked region with the data behind it, rather than have the masked region disappear from the image, as frequently I am not sure whether to mask features or not and need to see what I have done and possibly reconsider.  In Mask Management there's an option to make the regions visible, but then I don't see the outline of mask, and the region also becomes unmasked in the data lineout.  I suggest to show by default just the outline of the masked region and also subtract that region from the lineout, but then in mask management allow the option to remove the masked regions from the image.  I did not like that the menu for creating a new masked region disappeared every time after I pushed 'OK'... I had to go back to the menu and find it again every time.... very tedious as TARDIS data frequently has dozens of artifacts.  I could not get the 'Mask ellipse/rectangle ' to work.  I can draw a black shape on the data but nothing seems to be getting masked, and the 'undo last selection' button doesn't remove the black shape.  It would be nice to have the option to rotate the rectangle or ellipse.  I was trying to use the pinhole mask to get rid of noise at the edge of the plates and so was making repeated adjustments to the pinhole radius, but every time I push ok the menu disappears and I have to re-open it and then same default values for radius and thickness show up, rather than the last ones I selected, so I have to try to remember my previous choice.

**Comment by psavery** [comment]
Amy, thank you for your feedback. This is very helpful. We certainly want to improve the process so it is as intuitive as it can be. I will look into these in more detail soon.

**Comment by psavery** [comment]
@amylazicki, we found out that matplotlib very recently removed a method we were using. That is why the rectangle/ellipse masking wasn't working, and this should be fixed in #1413.

This would also cause an issue with removing TARDIS boundaries in the LLNL import tool (I'm not sure if this is documented anywhere). That should hopefully be fixed by #1413 too...

**Reply by amylazicki** [reply]
That's great!  thank you so much!  I will be having another go at this this Friday.

**Comment by psavery** [comment]
Here are some other responses:

```
In hand-drawn masks: what does right-clicking do?... what do you mean by a new line? Nothing
seems to happen for me with a right click. I can outline a region with left clicks and it's successfully 
masked (most of the time... sometimes for some reason I can't find, both plates just disappear and
then I just have to open the mask management menu and delete that mask). Sometimes the left clicks 
appear as blue dots, sometimes as orange or green or red... what does this mean?
```

The hand-drawn mask tool was built off of another tool that we had called a "line builder"... actually Joel designed the "line builder". The way it works is you start by left-clicking to add points (you are creating the region to be masked). When you are finished with one mask region, right-click one time to start a new region, and then left-click again to add points for this new region. You will see that the points are a different color, and a different color means you are creating a mask for a different region. So the example image below has four regions that will be masked (each one has different colored points).

![four_regions](https://user-images.githubusercontent.com/9558430/234702907-dc785aec-1778-4425-9b40-a99e8a6671da.png)

I agree that we need to clarify this. I think in that dialog, we should be clearer and say things like "Add points around a region to mask with left-click. Start a new masked region with a single right-click." I think @bnmajor will do that, and she will be adding more documentation on the masking soon - that will hopefully help a lot. However, feel free to give us ideas for making it more intuitive, if you think there are better ways to do this...

```
I would sort of like to be able to see the outline of the masked region with the data behind it,
rather than have the masked region disappear from the image, as frequently I am not sure
whether to mask features or not and need to see what I have done and possibly reconsider. In
Mask Management there's an option to make the regions visible, but then I don't see the outline
of mask, and the region also becomes unmasked in the data lineout. I suggest to show by default
just the outline of the masked region and also subtract that region from the lineout, but then in
mask management allow the option to remove the masked regions from the image.
```

I agree! This would make keeping track of masks a lot easier! I'll put up an issue.

```
I did not like that the menu for creating a new masked region disappeared every time after I
pushed 'OK'... I had to go back to the menu and find it again every time.... very tedious as TARDIS
data frequently has dozens of artifacts.
```

So, maybe the explanation above for hand-drawn masks helps with this? You can add multiple masks
before clicking "OK". Also, when #1413 is in, you can do the ellipse/rectangle masks too (where you
can also add multiple masks without clicking "OK").

```
It would be nice to have the option to rotate the rectangle or ellipse.
```

I'll put up an issue for this too!

```
I was trying to use the pinhole mask to get rid of noise at the edge of the plates and so was
making repeated adjustments to the pinhole radius, but every time I push ok the menu disappears
and I have to re-open it and then same default values for radius and thickness show up, rather than
the last ones I selected, so I have to try to remember my previous choice.
```

This should be an easy fix - we can make it remember the last setting. I'll put up an issue as well and
fix it very soon!

**Reply by amylazicki** [reply]
thank you!  I like the new descriptions in the hand-drawn mask menu

**Comment by psavery** [comment]
@amylazicki I put up a pull request to make the pinhole mask more interactive here: #1418. It also saves the pinhole mask settings that are used.

**Reply by amylazicki** [reply]
This is great!  Works well.

**Comment by amylazicki** [comment]
I wonder if you could provide some options to adjust the parameters for the 'apply erosion' feature... it misses a few pixels for me sometimes, and in some cases there are some extra background features toward the rind due to shadows or fluorescence off the fluorescence shield fixturing which I would like to mask out easily (see example below ... the high background regions at low angle that follow the curvature of the image plate).  Or maybe there is a better way to do this?:
![image](https://user-images.githubusercontent.com/130735505/235978273-6c150b95-db77-4658-8e53-e236be91f3fa.png)

**Reply by psavery** [reply]
The code where we are applying the erosion is [here](https://github.com/HEXRD/hexrdgui/blob/a3b31d1157ffad204f3ae9ccc70723c5381c0b04/hexrd/ui/calibration/polarview.py#L392-L399).

It looks like it creates the structure for the erosion from the SNIP options (specifically, the SNIP width and number of iterations). You can modify the erosion by modifying those. It looks like there are extra parameters we could add options for, too, if they would be helpful. The docs for the `binary_erosion` function are [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_erosion.html#scipy.ndimage.binary_erosion).

@saransh13 do you have any suggestions?
